src/main/java/com/coveo/pushapiclient/ApiUrl.java

```diff
@@ -0,0 +1,107 @@
+package com.coveo.pushapiclient;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Private util class to extract dynamic parts from a API URL
+ *
+ * @See https://docs.coveo.com/en/1546#push-api-url
+ *      https://docs.coveo.com/en/3295#stream-api-url
+ */
+class ApiUrl {
+    private final String organizationId;
+    private final String sourceId;
+    private final PlatformUrl platformUrl;
+
+    public ApiUrl(URL sourceUrl) throws MalformedURLException {
+        List<String> identifiers = this.extractIdentifiers(sourceUrl);
+        this.organizationId = identifiers.get(0);
+        this.sourceId = identifiers.get(1);
+        this.platformUrl = this.extractPlatformUrl(sourceUrl);
+    }
+
+    public String getOrganizationId() {
+        return this.organizationId;
+    }
+
+    public String getSourceId() {
+        return this.sourceId;
+    }
+
+    public PlatformUrl getPlatformUrl() {
+        return this.platformUrl;
+    }
+
+    private List<String> extractIdentifiers(URL sourceUrl) throws MalformedURLException {
+        String host = sourceUrl.getPath();
+        Pattern pattern = Pattern.compile("/push/v1/organizations/([^/]+)/sources/([^/]+)");
+        Matcher matcher = pattern.matcher(host);
```

Comment:
Nitpick, but in this case, you're not reading the host, you're reading the path.

```suggestion
        String path = sourceUrl.getPath();
        Pattern pattern = Pattern.compile("/push/v1/organizations/([^/]+)/sources/([^/]+)");
        Matcher matcher = pattern.matcher(path);
```